drivers: ieee802154: support Key Identifier Mode > 1

jciupis opened the pull request:
IEEE 802.15.4-2020 defines four possible values for Key Identifier Mode field of the Auxiliary Security Header. The current ieee802154 driver API only supports two of them: b00 and b01. This commit adds support for the two remaining Key Identifier Mode values. It's done by replacing a field that can only hold Key Index into a field that can holds a pointer to the entire Key Identifier field.

See IEEE 802.15.4-2020, sections 9.4.2.3 and 9.4.4 for further reference.

Commit:
drivers: ieee802154: support Key Identifier Mode > 1

IEEE 802.15.4-2020 defines four possible values for Key Identifier Mode
field of the Auxiliary Security Header. The current ieee802154 driver
API only supports two of them: b00 and b01. This commit adds support for
the two remaining Key Identifier Mode values. It's done by replacing a
field that can only hold Key Index into a field that can holds a pointer
to the entire Key Identifier field.

See IEEE 802.15.4-2020, sections 9.4.2.3 and 9.4.4 for further reference.

Signed-off-by: Jędrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

jciupis commented:
FYI: @ankuns

ankuns commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -781,7 +781,7 @@ static void nrf5_config_mac_keys(struct ieee802154_key *mac_keys)
 			.value.p_cleartext_key = keys->key_value,
 			.id.mode = keys->key_id_mode,
-			.id.p_key_id = &(keys->key_index),
+			.id.p_key_id = keys->key_id,
```
Is it enough? The meaning of what `key.id.p_key_id` points to is changed. I can point to multi-byte array.
What about line 796? `stored_ids` was also prepared for storing old `key_index` of `uint8_t` type.
Also line 798 seems now incorrect. stored_key_ids[i].p_key_id will point now to one byte in stored_ids[i] no matter what's the size of real key_id.

jciupis replied:
That's a valid concern, thanks for catching it. I put `DNM` label on the PR and will come back with a solution next week.

ankuns replied:
All this issues come from the fact that there is no function `nrf_802154_security_key_remove_all(void)`.
To be able to remove all keys we stored we need to track the key ids we have written.

Commit:
modules: hal_nordic: nrf_802154: remove magic number

This commit replaces a magic number in the definition of serialization
ring buffer length with a macro provided by the 802.15.4 driver.

Signed-off-by: Jędrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

edmont commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -781,36 +781,24 @@ static void nrf5_iface_init(struct net_if *iface)
-	static nrf_802154_key_id_t stored_key_ids[NRF_802154_SECURITY_KEY_STORAGE_SIZE];
-	static uint8_t stored_ids[NRF_802154_SECURITY_KEY_STORAGE_SIZE];
-	uint8_t i;
+	nrf_802154_security_key_remove_all();
```
Much better approach, thanks.

rlubos commented on west.yml:
```diff
@@ -183,7 +183,7 @@ manifest:
         - hal
     - name: hal_nordic
-      revision: 56e0b052dff311c2f8eb08c6804e60fc79feb56f
+      revision: pull/153/head
```
`6efb0fdef01737fac03ea500b28b6c4d12248af4`

jciupis commented:
The regression mentioned in the previous comment was caused by the assumption I made that the OpenThread stack always configures ieee802154 mac keys with Key Id Mode == 1. That was wrong because RCP uses Key Id Mode == 0 when resetting. I corrected the implementation and resolved conflicts that came up since last week.

jciupis commented:
FYI: @canisLupus1313, @edmont
